Add clamping to fp32->fp8e4m3 conversion as per issue #154028

KhonaB opened the pull request:
Fixes #154028

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @xmfan

KhonaB committed:
Add clamping to fp32->fp8e4m3fn conversion

KhonaB committed:
Add clamping to fp32->fp8e4m3fn conversion

KhonaB reviewed (commented):
Hopefully fixes #154028

KhonaB committed:
Switch to bitwise operations

KhonaB committed:
Switch to bitwise operations

KhonaB committed:
Analogous changes to vec512

KhonaB committed:
Analogous changes to vec512

KhonaB commented:
Passes all 33 tests listed by VS under "e4m3"--some other tests fail or get stuck, but those appear to happen regardless of whether the changes are made or not. It's possible I've done something completely wrong, though, not totally clear on how torchgen works, so a sanity check would be appreciated.

KhonaB commented:
Ah. Didn't see that requirement, first PR. Sorry.

I believe this is not user facing as the fp8 datatype is not intended for use as individual variables but rather as an internal optimization to make training/inference faster, and this change should just be a minor fix to it. Failing that, though, I'm not sure which release notes category to put it in, I don't know which one the floatx module belongs to (if it belongs to one at all).

@pytorchbot label "module: floatx (formerly float8)"
@pytorchbot label "topic: not user facing"

janeyx99 commented:
@KhonaB I've kicked off the CI for your PR, the classification updates every 15m or so if there are new signals it will update!

KhonaB committed:
Linter issues in Float8_e4m3fn.h

KhonaB committed:
Linter issues in Float8_e4m3fn.h

KhonaB committed:
Linting Float8_e4m3fn.h

KhonaB committed:
Linting Float8_e4m3fn.h

KhonaB committed:
Missing semicolon vec512_float8.h

KhonaB committed:
Missing semicolon vec512_float8.h

KhonaB committed:
Logic error with overflow/nan mask in vec512_float8.h

KhonaB committed:
Logic error with overflow/nan mask in vec512_float8.h

KhonaB commented:
@pytorchbot drci

leslie-fang-intel reviewed (commented):
Thanks for the PR! The changes in `vec512_float8.h` make sense to me as to match the scalar version's behavior. The core question now is whether we should follow the specification outlined in [arxiv.org/pdf/2209.05433](https://arxiv.org/pdf/2209.05433) or default to using saturated casts in PyTorch eager mode. There seems already has some discussion in https://github.com/pytorch/pytorch/issues/154028 cc @vkuzo

KhonaB commented:
alright, running the full test suite after running some (mostly preexisting) fp8 tests and catching an edge case on handling values that should have been clamped but instead were rounding/truncating to 480 (NaN in fp8e4m3fn), its taking a while presumably because I do not have a compatible accelerator (which should be fine, these were all CPU emulations of accelerator behavior) but once its complete I'll commit the changes to the PR.

Separately, I saw there are other types that are currently overflowing to NaN with specification from the same paper and tested together, specifically fp8e5m2fn and fp8e4m3fnuz. I could change those as well here, or in a separate PR once this is approved? Regardless, I had to change the expected behavior in the test because the prior tests were expecting non-saturating behavior from all fp8 fn datatypes, which seems to contradict the specification.

KhonaB committed:
Floatx tests cases updated, bugfix on 464.0<x<480.0 rounding to NaN

KhonaB committed:
Merge branch 'fp8e4m3fn-clamping' into fp8fix

KhonaB committed:
Merge bugfixes and test changes from KhonaB/fp8fix

Branch merger

sladyn98 reviewed (changes requested):
Code is failing some lint and tests. Kindly fix and request review.

KhonaB committed:
Linter fix

KhonaB committed:
Initial base of clamping branch was tip of pytorch master, bad idea
Fix via merge of remote-tracking branch 'upstream/2.9.1' into fp8e4m3fn-clamping
Fixes test failure.
